Introduce Freezing fee

omerfirmak opened the pull request:
Freezing TXs now have to include SlashAmountPenatly amount of fee
for each Freeze output they have. This amount will be refunded to them
if they are not slashed for the duration of their validation cycle.

Part of #2554

omerfirmak committed:
Introduce Freezing fee

Freezing TXs now have to include SlashAmountPenatly amount of fee
for each Freeze output they have. This amount will be refunded to them
if they are not slashed for the duration of their validation cycle.

omerfirmak committed:
Fee: Rename storeBlockFees to storeValidatedBlockFees

to emphasize that the function assumes a valid block.

omerfirmak committed:
Ledger: Disallow indirect EnrollmentManager access from Ledger

only allowed if you have a TestLedger instance.

omerfirmak committed:
Enrollment: Require PenaltyAmount to be still controlled by the network

Except the genesis enrollments

hewison-chris commented on source/agora/consensus/Fee.d:
```diff
@@ -20,6 +20,7 @@ module agora.consensus.Fee;
 import agora.common.Types;
 import agora.common.Amount;
 import agora.common.Ensure;
+import agora.common.Set;
```
Should be one line below.

hewison-chris commented on source/agora/consensus/Fee.d:
```diff
@@ -364,9 +395,13 @@ public class FeeManager
 
+        // Subtract the freeze fees
+        if (!total_payout.sub(this.getBlockFreezeFees(height)))
+            assert(0, "getCommonsBudgetPayout: Failed to sub freeze fees");
```
```suggestion
            assert(0, "getCommonsBudgetPayout: Failed to subtract freeze fees");
```

hewison-chris commented on source/agora/consensus/Fee.d:
```diff
@@ -389,21 +424,21 @@ public class FeeManager
+        foreach (tx; block.txs)
+        foreach (idx, o; tx.outputs)
+            if (o.type == OutputType.Freeze)
+                this.withhold_freeze_fees.put(UTXO.getHash(tx.hashFull(), idx));
```
```suggestion
            foreach (idx, o; tx.outputs)
                if (o.type == OutputType.Freeze)
                    this.withhold_freeze_fees.put(UTXO.getHash(tx.hashFull(), idx));
```

hewison-chris commented on source/agora/consensus/Fee.d:
```diff
@@ -572,7 +639,8 @@ public class FeeManager
         Transaction gen_tx = Transaction(
-            [ Output(Amount(20_000_000L), WK.Keys.NODE2.address) ]);
+            [ Output(Amount(20_000_000L), WK.Keys.NODE2.address),
+              Output(Amount.MinFreezeAmount + 10_000.coins, WK.Keys.NODE2.address)]);
```
```suggestion
              Output(Amount.MinFreezeAmount + 10_000.coins, WK.Keys.NODE2.address) ]);
```

hewison-chris commented on source/agora/consensus/Fee.d:
```diff
@@ -641,8 +714,8 @@ public class FeeManager
 
-        // None wasted, none created
-        assert(tot_fee == Amount(0));
+        // None wasted, none created (Freeze fee is withhold)
```
```suggestion
        // None wasted, none created (Freeze fee is withheld)
```

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -139,7 +133,7 @@ public class Ledger
         Engine engine, UTXOCache utxo_set, IBlockStorage storage,
         EnrollmentManager enroll_man, TransactionPool pool,
-        ManagedDatabase stateDB, Clock clock,
+        FeeManager fee_man, ManagedDatabase stateDB, Clock clock,
```
Maybe `stateDB` is unused now?

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -363,29 +357,22 @@ public class Ledger
             return "Fee rate is lower than this node's configured relative threshold (min_fee_pct)";
         if (!this.isAcceptableDoubleSpent(tx, double_spent_threshold_pct))
             return "Double spend comes with a less-than-acceptable fee increase";
-
```
Normally we include this blank line before last `return`

hewison-chris commented on source/agora/consensus/pool/Enrollment.d:
```diff
@@ -19,6 +19,7 @@ module agora.consensus.pool.Enrollment;
 import agora.common.ManagedDatabase;
 import agora.common.Types;
 import agora.crypto.Key;
+import agora.consensus.Fee;
```
Import is not sorted.

hewison-chris commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -18,6 +18,7 @@ module agora.consensus.state.ValidatorSet;
 import agora.common.Amount;
 import agora.common.ManagedDatabase;
 import agora.common.Types;
+import agora.consensus.Fee;
```
Not sorted.

hewison-chris commented on source/agora/consensus/validation/Transaction.d:
```diff
@@ -479,13 +480,40 @@ unittest
+        );
+        secondTx.inputs[0].unlock = signUnlock(key_pairs[0], secondTx);
+        // Second Transaction is invalid.
+        assert(!secondTx.isValid(engine, storage.getUTXOFinder(), Height(0), checker));
```
Would be nice to then update `secondTx` to be valid and assert it is.

omerfirmak replied:
There are other tests checking for valid freeze txs.

hewison-chris replied:
I was meaning to verify that this assert is false for the right reason within this test.

hewison-chris replied:
I pulled the branch and had a better look and I now see the test before is almost the same but with valid Amount to include enough fees.
The tests could perhaps be a bit DRYer to make it easier to see the difference between each but I think you have the different cases covered here.

hewison-chris commented on source/agora/test/ManyValidators.d:
```diff
@@ -35,17 +35,8 @@ void manyValidators (size_t validators)
-            .each!(tx => network.clients[0].postTransaction(tx));
-    }
+    // prepare frozen outputs for the outsider validator to enroll
+    network.postAndEnsureTxInPool(network.freezeUTXO(iota(GenesisValidators, GenesisValidators + conf.outsider_validators)));
```
:+1:

hewison-chris commented on source/agora/utils/TxBuilder.d:
```diff
@@ -353,6 +354,11 @@ public struct TxBuilder
             o.type = outputs_type;
 
         auto total_fees = this.minFees();
+        auto out_count = this.data.outputs.count!(o => o.type == OutputType.Freeze);
```
```suggestion
        auto freeze_outputs = this.data.outputs.count!(o => o.type == OutputType.Freeze);
```

hewison-chris reviewed (changes requested):
Nice work, looks good just some small changes suggested.

omerfirmak committed:
ValidatorSet: Return slashed_height in EnrollmentState

This will be usefull when determining if we are supposed to pay
the penalty the commons budget

omerfirmak committed:
Fee: Pay SlashPenalty to Commons Budget

in the case where the validator is slashed

omerfirmak committed:
Transaction: Allow slash penalty to be spent while unfreezing UTXOs

omerfirmak commented:
There are bunch of races between enrollments and spending the penalty deposit with changes in last commit, but the rest of the changes are stable enough to be reviewed.

omerfirmak committed:
Ledger: Prevent stakes to be spent on enrollment block

Stakes are locked for the enrollment duration, but could be
spent on the same block.

omerfirmak committed:
Ledger: Lock enrolled stakes for the duration of the enrollment

omerfirmak commented:
Fixes #2622 and #2554

Geod24 commented on source/agora/consensus/Fee.d:
```diff
@@ -279,6 +294,14 @@ public class FeeManager
+    public Amount getFreezeFee (in Transaction tx) pure nothrow @safe @nogc
+    {
+        Amount freeze_fee = this.params.SlashPenaltyAmount;
+        freeze_fee.mul(tx.outputs.count!(output => output.type == OutputType.Freeze));
```
Missing check

omerfirmak replied:
Callers do the checks whenever needed

Geod24 reviewed (commented):
First commit:
- `SlashAmountPenatly` (commit message): Typo
- I still need to process the details, but the idea looks pretty good;
- I don't see the genesis block being modified though ?

- 2nd commit: 👌 
- 3rd commit: ❤️ ❤️ ❤️ ❤️ 
- 4th commit: 👌 
-5th to 7th commit: Lots of change... And not doing it for Genesis is a code smell. Still need to process it though.
8th, 9th: Not sure this is the right way to go. Let's have a quick chat.

omerfirmak commented:
> 
> 
> First commit:
> 
>     * `SlashAmountPenatly` (commit message): Typo
> 
>     * I still need to process the details, but the idea looks pretty good;
> 
>     * I don't see the genesis block being modified though ?
> 
>     * 2nd commit: 👌
> 
>     * 3rd commit: ❤️ ❤️ ❤️ ❤️
> 
>     * 4th commit: 👌
>       -5th to 7th commit: Lots of change... And not doing it for Genesis is a code smell. Still need to process it though.
>       8th, 9th: Not sure this is the right way to go. Let's have a quick chat.

The problem with genesis is that there are no inputs for the freeze outputs. So, checking for "fees" in genesis does not really make sense. We could deduct the penalty amount from their frozen outputs tho, just to keep the supply the same.

omerfirmak commented:
Rebased and in desperate need of some reviews.

omerfirmak commented on source/agora/consensus/Fee.d:
```diff
@@ -249,6 +259,11 @@ public class FeeManager
             return "Fee: Transaction size overflows fee cap";
+
+        auto freeze_fee = this.getFreezeFee(tx);
+        if (!freeze_fee.isValid() || !minimumFee.add(freeze_fee))
```
Since `tx` is user controlled, `getFreezeFee` is allowed to return invalid values.

omerfirmak commented on source/agora/consensus/Fee.d:
```diff
@@ -364,9 +387,13 @@ public class FeeManager
             assert(0, "getCommonsBudgetPayout: Failed to add total fees");
 
+        // Subtract the freeze fees
+        if (!total_payout.sub(this.getBlockFreezeFees(height)))
```
Freeze fees are not included in the payouts unless the validator is slashed.

omerfirmak commented on source/agora/consensus/Fee.d:
```diff
@@ -500,8 +540,12 @@ public class FeeManager
-            rate = tx.getFee(peekUTXO);
+            auto freeze_fee = this.getFreezeFee(tx);
+            ensure(freeze_fee.isValid(), "Can't calculate freeze fee");
+            rate = tx.getFee(peekUTXO) - freeze_fee;
```
Freeze fees are excluded from rate calculations or they highly skew the minimum fee for TX pool. acceptance

hewison-chris replied:
Maybe it is better that freeze `tx`s are always chosen before other `tx`s to enable enrolments.

hewison-chris replied:
I think I understand now. This is to prevent `tx` rejection due to  https://github.com/bosagora/agora/blob/1f67ab377ac085fb70ebe81ec6ff441421a8d624/source/agora/node/Config.d#L240-L266

omerfirmak replied:
Exactly:)

omerfirmak commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -48,6 +48,9 @@ public struct EnrollmentState
     PreImageInfo preimage;
+
+    /// The Height the validator was slashed at
+    Height slashed_height;
```
see the commit message

omerfirmak commented on source/agora/consensus/Ledger.d:
```diff
@@ -677,6 +677,18 @@ public class Ledger
+        UTXO utxo_val;
+        if (!this.peekUTXO(utxo, utxo_val))
+            return 0.coins;
+        return this.params.SlashPenaltyAmount;
```
TX validation ensures that freezing fee was paid for each frozen output. 

Freezing fee will still be controlled by the network only if the frozen output is still unspent and validator was never slashed while enrolled with that UTXO. Once a validator is slashed, freezing fee is paid to the commons budget as penalty in the next payout block.

If validator finishes the cycle without getting slashed it is allowed to spend the frozen utxo along with the freezing fee it paid in the freezing TX.

omerfirmak commented on source/agora/consensus/validation/Enrollment.d:
```diff
@@ -93,6 +95,9 @@ public string isInvalidReason (in Enrollment enrollment,
         return Message;
     }
 
+    if (hasPenaltyDeposit(enrollment.utxo_key) == 0.coins)
```
This function will return 0.coins if the utxo is spent or that stake was slashed. So even if the frozen UTXO is still in the pool, validator wont be able to reenroll with the same stake.

omerfirmak commented on source/agora/consensus/validation/Transaction.d:
```diff
@@ -94,10 +100,12 @@ public string isInvalidReason (
     {
         if (!findUTXO(input.utxo, utxo_value))
             return "Transaction: Input ref not in UTXO";
+        if (!sum_unspent.add(hasPenaltyDeposit(input.utxo)))
```
If the input is a freeze output and was not slashed, we allow TX to spend the penalty deposit. But beware, we dont force them to spend it. if they dont add the penalty deposit to the outputs, that deposit is lost forever, essentially burned.

hewison-chris replied:
Would it be fairer if it was refund by default?

omerfirmak replied:
This was the simplest, yet works-perfectly-fine solution I could think of :) Since we probably want the refund to also be subject to the melting period we apply to frozen stakes, refunding them in a Coinbase TX was not my first choice.

omerfirmak commented on source/agora/consensus/Ledger.d:
```diff
@@ -535,6 +535,7 @@ public class Ledger
                 log.fatal("Validated block: {}", block);
                 assert(0);
             }
+            this.utxo_set.updateUTXOLock(enrollment.utxo_key, block.header.height + this.params.ValidatorCycle);
```
we have to lock the UTXO for the enrollment duration. if not, validator can spend the stake and the penalty deposit. Even tho the outputs are locked (ie melting), we no longer can slash them since we lose track of the stake.

omerfirmak commented on source/agora/consensus/Ledger.d:
```diff
@@ -1489,9 +1516,10 @@ public class ValidatingLedger : Ledger
-        data.enrolls = this.getCandidateEnrollments(next_height);
-        data.missing_validators = this.getCandidateMissingValidators(next_height);
-        data.tx_set = this.getCandidateTransactions(next_height, max_txs);
+        auto utxo_finder = this.utxo_set.getUTXOFinder();
```
by using the same utxo_finder with double spend protection, we simply force a UTXO to be referenced only once in a block, either in a enrollment or a TX.

omerfirmak reviewed (commented):
Self reviewing, hoping to lower the bar of entry for team reviews:)

hewison-chris commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -195,7 +196,8 @@ public class EnrollmentManager
     public bool addEnrollment (in Enrollment enroll,
-        in PublicKey pubkey, in Height height, scope UTXOFinder finder)
+        in PublicKey pubkey, in Height height, scope UTXOFinder finder,
+        scope HasPenaltyDeposit hasPenaltyDeposit)
```
The param is missing from function DOC.

hewison-chris commented on source/agora/common/Types.d:
```diff
@@ -179,3 +180,6 @@ unittest
 }
+
+/// Delegate type to query the penalty deposit of a utxo
+public alias HasPenaltyDeposit = Amount delegate (Hash utxo) @safe nothrow;
```
```suggestion
public alias PenaltyDeposited = Amount delegate (Hash utxo) @safe nothrow;
```

hewison-chris replied:
I guess it was a boolean at some point.

omerfirmak replied:
Phew, I cried a little doing that:)

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -1334,7 +1347,28 @@ public class Ledger
+
+    /***************************************************************************
+
+        Add a enrollment data to the enrollment pool
```
```suggestion
        Add an enrollment data to the enrollment pool
```

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -2358,3 +2392,20 @@ unittest
+    auto freeze_tx = GenesisBlock.txs.find!(tx => tx.isFreeze).front();
+    auto melting_tx = TxBuilder(freeze_tx, 0).sign();
+
+    // enrolled stake cant be spent
```
```suggestion
    // enrolled stake can't be spent
```

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -2353,7 +2339,8 @@ unittest
             // Payout block should pay the CommonsBudget + all validators (excluding slashed validators)
             assert(cb_txs[0].outputs.length == 1 + genesis_validator_keys.length - skip_indexes.length);
-            assert(cb_txs[0].outputs.map!(o => o.value).reduce!((a,b) => a + b) ==  next_payout_total);
+            assert(cb_txs[0].outputs.map!(o => o.value).reduce!((a,b) => a + b) == next_payout_total[0]);
```
Looks like you fixed a test bug here. 👍

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -2420,3 +2421,19 @@ unittest
+    auto freeze_tx = GenesisBlock.txs.find!(tx => tx.isFreeze).front();
+    auto melting_tx = TxBuilder(freeze_tx, 0).sign();
+
+    // enrolled stake cant be spent
```
```suggestion
    // enrolled stake can't be spent
```

hewison-chris commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -236,15 +236,16 @@ public class EnrollmentManager
         Params:
             height = block height intended for the enrollments
-            peekUTXO = An `UTXOFinder` without replay-protection
+            peekUTOX = An `UTXOFinder` without replay-protection
```
```suggestion
            peekUTXO = An `UTXOFinder` without replay-protection
```

hewison-chris reviewed (changes requested):
Thanks for the self review. Just some requests, nothing major.

hewison-chris reviewed (approved):
LGTM

omerfirmak commented:
Problem with our current approach is obviously visible in testnet, bring up the wallet for VAL2 and you can see that it still thinks it has a frozen UTXO. But it was slashed, so that UTXO is no longer available and we have no record the refund UTXO.

hewison-chris commented:
Needs a rebase.

omerfirmak commented:
> Needs a rebase.

Rebased

omerfirmak commented:
I merged some new code in to https://github.com/bosagora/agora/pull/2597/commits/e7a6f4758f0e62b3cf798a732f300284211b7c23 , please take a look.

Melting TX is now forced to spend atleast some part of the penalty deposit.

hewison-chris commented:
> I merged some new code in to [e7a6f47](https://github.com/bosagora/agora/commit/e7a6f4758f0e62b3cf798a732f300284211b7c23) , please take a look.
> 
> Melting TX is now forced to spend atleast some part of the penalty deposit.

I think this is better now. Will re run ci as a couple failed but probably not related.
